File: backend/services/persistence_manager.py
"""
Persistence Manager - Unified interface for world initialization and sync.

This module provides a single point of access for:
1. World initialization (syncing filesystem content to database)
2. Creating locations (requires both filesystem and database)
3. Exporting database state back to filesystem (for backups)

Runtime state operations (stats, inventory, location changes) go directly
through crud.game.* to the database only.
"""

import json
import logging
from typing import Optional

import crud
import schemas
from sqlalchemy.ext.asyncio import AsyncSession

from services.location_service import LocationService
from services.player_service import PlayerService
from services.world_service import WorldService

logger = logging.getLogger("PersistenceManager")


class PersistenceManager:
    """
    Manages synchronized persistence for world initialization and export.

    This class handles:
    - Initial sync from filesystem to database (during onboarding→active transition)
    - Creating locations (requires both FS + DB)
    - Exporting database state to filesystem (for backup/portability)

    Runtime state changes (stats, inventory, travel) should use crud.game.* directly.
    """

    def __init__(self, db: AsyncSession, world_id: int, world_name: str):
        """
        Initialize the persistence manager.

        Args:
            db: Async database session
            world_id: Database ID of the world
            world_name: Filesystem name of the world (used for WorldService)
        """
        self.db = db
        self.world_id = world_id
        self.world_name = world_name

    # =========================================================================
    # Initialization Methods (Filesystem → Database)
    # =========================================================================

    async def create_location(
        self,
        name: str,
        display_name: str,
        description: str,
        position: tuple[int, int],
        adjacent_hints: Optional[list[str]] = None,
        is_starting: bool = False,
        agents: Optional[list[str]] = None,
    ) -> int:
        """
        Create a location in both filesystem and database.

        This is used during world seeding when new locations are created.

        Args:
            name: Internal location name (used as key/path)
            display_name: Human-readable display name
            description: Location description
            position: (x, y) position on the map
            adjacent_hints: List of adjacent location names (filesystem)
            is_starting: Whether this is the starting location
            agents: List of agent names at this location

        Returns:
            Database ID of the created location
        """
        # 1. Create in filesystem (source of truth)
        LocationService.create_location(
            self.world_name,
            name,
            display_name,
            description,
            position,
            adjacent=adjacent_hints,
        )
        logger.info(f"Created location '{name}' in filesystem")

        # 2. Create in database (includes room creation for messages)
        location_create = schemas.LocationCreate(
            name=name,
            display_name=display_name,
            description=description,
            position_x=position[0],
            position_y=position[1],
            adjacent_to=None,  # Will be linked later if needed
            is_discovered=True,
        )
        db_location = await crud.create_location(self.db, self.world_id, location_create)
        logger.info(f"Created location '{name}' in database (id={db_location.id}, room_id={db_location.room_id})")

        # 3. Store room mapping in _state.json (FS-first architecture)
        if db_location.room_id:
            room_key = LocationService.location_to_room_key(name)
            LocationService.set_room_mapping(
                world_name=self.world_name,
                room_key=room_key,
                db_room_id=db_location.room_id,
                agents=agents or [],
            )
            logger.info(f"Stored room mapping: {room_key} -> room_id={db_location.room_id}")

        # 4. If starting location, set as current location and room
        if is_starting:
            await crud.set_current_location(self.db, self.world_id, db_location.id)
            room_key = LocationService.location_to_room_key(name)
            LocationService.set_current_room(self.world_name, room_key)
            logger.info(f"Set '{name}' as current location and room")

        return db_location.id

    async def sync_player_state_from_filesystem(self) -> None:
        """
        Sync player state from filesystem to database.

        This is useful after WorldSeedManager runs, to ensure database
        reflects the filesystem state during onboarding→active transition.
        """
        # Load state from filesystem
        fs_state = PlayerService.load_player_state(self.world_name)
        if not fs_state:
            logger.warning(f"No player state found in filesystem for world '{self.world_name}'")
            return

        # Get database player state
        db_state = await crud.get_player_state(self.db, self.world_id)
        if not db_state:
            logger.warning(f"No player state found in database for world_id={self.world_id}")
            return

        # Sync stats
        if fs_state.stats:
            await crud.initialize_player_stats(
                self.db,
                self.world_id,
                {"stats": [{"name": k, "default": v} for k, v in fs_state.stats.items()]},
                initial_stats=fs_state.stats,
            )
            logger.info(f"Synced {len(fs_state.stats)} stats to database")

        # Sync inventory
        if fs_state.inventory:
            for item in fs_state.inventory:
                await crud.add_inventory_item(
                    self.db,
                    self.world_id,
                    schemas.InventoryItem(
                        id=item.get("item_id", item.get("id", "")),
                        name=item.get("name", ""),
                        description=item.get("description", ""),
                        quantity=item.get("quantity", 1),
                        properties=item.get("properties"),
                    ),
                )
            logger.info(f"Synced {len(fs_state.inventory)} inventory items to database")

        # Sync current location (need to find or create location)
        if fs_state.current_location:
            locations = await crud.get_locations(self.db, self.world_id)
            location = next(
                (loc for loc in locations if loc.name == fs_state.current_location),
                None,
            )

            if not location:
                # Location doesn't exist in database - create it from filesystem
                location = await self._create_location_from_filesystem(fs_state.current_location)

            if location:
                await crud.set_current_location(self.db, self.world_id, location.id)
                logger.info(f"Synced current_location to '{fs_state.current_location}' (id={location.id})")
            else:
                logger.warning(f"Could not create location '{fs_state.current_location}' from filesystem")

    async def _create_location_from_filesystem(self, location_name: str):
        """
        Create a location in the database from filesystem data.

        Args:
            location_name: The location name (directory name in filesystem)

        Returns:
            The created Location model, or None if failed
        """
        try:
            # Load location data from filesystem using LocationService
            loc_config = LocationService.load_location(self.world_name, location_name)
            if not loc_config:
                logger.warning(f"Location '{location_name}' not found in filesystem")
                return None

            # Check for existing room mapping to preserve agents added during onboarding
            room_key = LocationService.location_to_room_key(location_name)
            existing_mapping = LocationService.get_room_mapping(self.world_name, room_key)
            existing_agents = existing_mapping.agents if existing_mapping else []

            # Create location in database
            position = loc_config.position if isinstance(loc_config.position, tuple) else (0, 0)
            location_create = schemas.LocationCreate(
                name=location_name,
                display_name=loc_config.display_name,
                description=loc_config.description or "",
                position_x=position[0],
                position_y=position[1],
                adjacent_to=None,
                is_discovered=loc_config.is_discovered,
            )

            db_location = await crud.create_location(self.db, self.world_id, location_create)
            logger.info(f"Created location '{location_name}' in database (id={db_location.id})")

            # Store room mapping in _state.json (preserve existing agents)
            if db_location.room_id:
                LocationService.set_room_mapping(
                    world_name=self.world_name,
                    room_key=room_key,
                    db_room_id=db_location.room_id,
                    agents=existing_agents,
                )
                logger.info(f"Stored room mapping: {room_key} -> room_id={db_location.room_id}")

                # Add existing agents to the database room
                if existing_agents:
                    added_count = await self._add_agents_to_room(db_location.room_id, existing_agents)
                    logger.info(f"Added {added_count} character agents to room {db_location.room_id}")

            return db_location

        except Exception as e:
            logger.error(f"Failed to create location '{location_name}' from filesystem: {e}")
            return None

    async def _add_agents_to_room(self, room_id: int, agent_names: list[str]) -> int:
        """
        Add agents by name to a room.

        Args:
            room_id: Database room ID
            agent_names: List of agent names to add

        Returns:
            Number of agents successfully added
        """
        from crud.room_agents import add_agent_to_room

        added = 0
        for agent_name in agent_names:
            # Look up agent by exact name
            agent = await crud.get_agent_by_name(self.db, agent_name)
            if agent:
                await add_agent_to_room(self.db, room_id, agent.id)
                added += 1
                logger.debug(f"Added agent '{agent_name}' to room {room_id}")
            else:
                logger.warning(f"Agent '{agent_name}' not found in world '{self.world_name}'")
        return added

    async def save_stat_definitions(self, stat_definitions: dict) -> None:
        """
        Save stat definitions to both database and filesystem.

        This is used during world seeding.

        Args:
            stat_definitions: Dict with "stats" list and optional "derived" list
        """
        # 1. Save to filesystem
        PlayerService.save_stat_definitions(self.world_name, stat_definitions)
        logger.info(f"Saved stat definitions to filesystem for world '{self.world_name}'")

        # 2. Save to database
        await crud.update_world(
            self.db,
            self.world_id,
            schemas.WorldUpdate(stat_definitions=stat_definitions),
        )
        logger.info(f"Saved stat definitions to database for world_id={self.world_id}")

    async def update_world_phase(self, phase: str) -> None:
        """
        Update world phase in both database and filesystem.

        Args:
            phase: New phase ("onboarding", "active", "ended")
        """
        # 1. Update filesystem
        config = WorldService.load_world_config(self.world_name)
        if config:
            config.phase = phase
            WorldService.save_world_config(self.world_name, config)
            logger.info(f"Set phase='{phase}' in filesystem")
        else:
            logger.warning(f"Could not load world config for '{self.world_name}' to update phase")

        # 2. Update database
        await crud.update_world(self.db, self.world_id, schemas.WorldUpdate(phase=phase))
        logger.info(f"Set phase='{phase}' in database")

    async def sync_stats(self, stats: dict[str, int]) -> None:
        """
        Initialize stats in database from filesystem state.

        Used during world seeding to populate initial stats.

        Args:
            stats: Dictionary of stat_name -> value
        """
        await crud.initialize_player_stats(
            self.db,
            self.world_id,
            {"stats": [{"name": k, "default": v} for k, v in stats.items()]},
            initial_stats=stats,
        )
        logger.info(f"Synced {len(stats)} stats to database")

    # =========================================================================
    # Export Methods (Database → Filesystem)
    # =========================================================================

    async def export_state_to_filesystem(self) -> None:
        """
        Export current database state to filesystem.

        This allows creating backups or portable world exports.
        Exports: player state (stats, inventory, location), locations.
        """
        # Get current state from database
        player_state = await crud.get_player_state(self.db, self.world_id)
        locations = await crud.get_locations(self.db, self.world_id)

        if not player_state:
            logger.warning(f"No player state to export for world_id={self.world_id}")
            return

        # Build filesystem player state
        fs_state = PlayerService.load_player_state(self.world_name)
        if not fs_state:
            # Create new state if it doesn't exist
            from domain.entities.world_models import PlayerState

            fs_state = PlayerState(
                current_location=None,
                turn_count=0,
                stats={},
                inventory=[],
                effects=[],
                recent_actions=[],
            )

        # Update from database
        fs_state.turn_count = player_state.turn_count

        # Parse stats from JSON
        if player_state.stats:
            fs_state.stats = json.loads(player_state.stats)

        # Parse inventory from JSON
        if player_state.inventory:
            fs_state.inventory = json.loads(player_state.inventory)

        # Set current location name
        if player_state.current_location:
            fs_state.current_location = player_state.current_location.name

        # Save to filesystem
        PlayerService.save_player_state(self.world_name, fs_state)
        logger.info(f"Exported player state to filesystem for world '{self.world_name}'")

        # Export location discovered status
        for location in locations:
            loc_config = LocationService.load_location(self.world_name, location.name)
            if loc_config and loc_config.is_discovered != location.is_discovered:
                # Update filesystem location with discovered status
                LocationService.update_location(
                    self.world_name,
                    location.name,
                    is_discovered=location.is_discovered,
                    label=location.label,
                )

        logger.info(f"Exported {len(locations)} locations to filesystem")

File: backend/src/services/persistence-manager.ts
// The bridge between a world on disk and a world in the database, for the few
// operations that cannot live on one side. Not a general write path: a runtime
// mutation added here would gain a second source of truth.

import { getSettings } from '../config/settings'
import { getAgentByName } from '../crud/agents'
import { createLocation, getLocations, type LocationCreate } from '../crud/locations'
import {
  addInventoryItem,
  getPlayerState,
  initializePlayerStats,
  setCurrentLocation,
} from '../crud/player-state'
import { addAgentToRoom } from '../crud/rooms'
import { updateWorld } from '../crud/worlds'
import type { Db } from '../db'
import type { Location, WorldPhase } from '../db/schema'
import type { InventoryEntry } from '../domain/player-rules'
// `worlds.stat_definitions`; StatDefinitionsFile is the whole `stats.json`.
import type { StatDefinitions as StatDefinitionBlob } from '../domain/player-rules'
import { getLogger } from '../infrastructure/logging/logger'
import { LocationStorage } from './location-storage'
import {
  DEFAULT_GAME_TIME,
  PlayerService,
  type PlayerState,
  type StatDefinitions as StatDefinitionsFile,
} from './player-service'
import { RoomMappingService } from './room-mapping'
import { MtimeCache, WorldService } from './world-service'

const logger = getLogger('PersistenceManager')

export type Position = [number, number]

export interface CreateLocationInput {
  /** Directory name, and the key every other layer refers to the place by. */
  name: string
  displayName: string
  description: string
  position: Position
  /** Neighbour *names*, written to the filesystem index only. */
  adjacentHints?: string[] | null
  isStarting?: boolean
  agents?: string[] | null
}

// The value is passed as `default` *and* as the initial value: without the
// definition entry the stat is never initialised at all.
function statDefinitionsFrom(stats: Record<string, number>): StatDefinitionBlob {
  return { stats: Object.entries(stats).map(([name, value]) => ({ name, default: value })) }
}

function blankPlayerState(): PlayerState {
  return {
    currentLocation: null,
    turnCount: 0,
    stats: {},
    inventory: [],
    effects: [],
    recentActions: [],
    // Must not be empty: the writer dumps it into `player.json`, and every
    // reader of that file expects a clock.
    gameTime: { ...DEFAULT_GAME_TIME },
    equipment: {},
    flags: {},
  }
}

export class PersistenceManager {
  private readonly worlds: WorldService
  private readonly players: PlayerService
  private readonly locations: LocationStorage
  private readonly rooms: RoomMappingService

  // `worldId` addresses the database half, `worldName` the filesystem half.
  constructor(
    private readonly db: Db,
    private readonly worldId: number,
    private readonly worldName: string,
    worldsDir: string = getSettings().paths.worldsDir,
  ) {
    // One cache across all three, so a write through any one is visible to the
    // others' next read.
    const cache = new MtimeCache()
    this.worlds = new WorldService(worldsDir, cache)
    this.players = new PlayerService(worldsDir, cache)
    this.locations = new LocationStorage(worldsDir, cache)
    this.rooms = new RoomMappingService(worldsDir)
  }

  /**
   * Create a location on both sides. Order is load-bearing: filesystem first
   * (source of truth), room mapping last (needs the allocated `room_id`).
   * `adjacentHints` reaches the filesystem only — the column takes ids, and the
   * caller calls `addAdjacentLocation` once both ends exist.
   */
  createLocation(input: CreateLocationInput): number {
    this.locations.createLocation(
      this.worldName,
      input.name,
      input.displayName,
      input.description,
      input.position,
      input.adjacentHints ?? [],
    )
    logger.info(`Created location '${input.name}' in filesystem`)

    const locationCreate: LocationCreate = {
      name: input.name,
      displayName: input.displayName,
      description: input.description,
      positionX: input.position[0],
      positionY: input.position[1],
      adjacentTo: null,
      isDiscovered: true,
    }
    const dbLocation = createLocation(this.db, this.worldId, locationCreate)
    logger.info(
      `Created location '${input.name}' in database (id=${dbLocation.id}, ` +
        `room_id=${String(dbLocation.roomId)})`,
    )

    const roomKey = RoomMappingService.locationToRoomKey(input.name)

    if (dbLocation.roomId) {
      this.rooms.setRoomMapping(this.worldName, roomKey, dbLocation.roomId, input.agents ?? [])
      logger.info(`Stored room mapping: ${roomKey} -> room_id=${dbLocation.roomId}`)
    }

    if (input.isStarting) {
      // The agents go in the mapping but not the database room: this path runs
      // during onboarding, before the characters have rows to add.
      setCurrentLocation(this.db, this.worldId, dbLocation.id)
      this.rooms.setCurrentRoom(this.worldName, roomKey)
      logger.info(`Set '${input.name}' as current location and room`)
    }

    return dbLocation.id
  }

  /**
   * Copy `player.json` into the database: the onboarding→active handover, and
   * polling's repair path. Empty sections are skipped rather than written as
   * empty, so this cannot wipe what the database has and the file does not.
   */
  syncPlayerStateFromFilesystem(): void {
    const fsState = this.players.loadPlayerState(this.worldName)
    if (!fsState) {
      logger.warning(`No player state found in filesystem for world '${this.worldName}'`)
      return
    }

    if (!getPlayerState(this.db, this.worldId)) {
      logger.warning(`No player state found in database for world_id=${this.worldId}`)
      return
    }

    const statNames = Object.keys(fsState.stats)
    if (statNames.length > 0) {
      initializePlayerStats(
        this.db,
        this.worldId,
        statDefinitionsFrom(fsState.stats),
        fsState.stats,
      )
      logger.info(`Synced ${statNames.length} stats to database`)
    }

    if (fsState.inventory.length > 0) {
      for (const item of fsState.inventory) {
        addInventoryItem(this.db, this.worldId, {
          // `player.json` stores the *reference* format, so name/description
          // are usually empty — the `items/` template holds them, unresolved here.
          id: String(item.item_id ?? item.id ?? ''),
          name: String(item.name ?? ''),
          description: item.description ?? '',
          quantity: item.quantity ?? 1,
          properties: item.properties ?? null,
        })
      }
      logger.info(`Synced ${fsState.inventory.length} inventory items to database`)
    }

    if (fsState.currentLocation) {
      const locations = getLocations(this.db, this.worldId)
      const location =
        locations.find((candidate) => candidate.name === fsState.currentLocation) ??
        this.createLocationFromFilesystem(fsState.currentLocation)

      if (location) {
        setCurrentLocation(this.db, this.worldId, location.id)
        logger.info(
          `Synced current_location to '${fsState.currentLocation}' (id=${location.id})`,
        )
      } else {
        logger.warning(
          `Could not create location '${fsState.currentLocation}' from filesystem`,
        )
      }
    }
  }

  /**
   * Materialise a database row for a location that only exists on disk. The new
   * `room_id` must be written back *with the existing agent list intact*, or the
   * characters in the starting location vanish once the world goes active.
   */
  private createLocationFromFilesystem(locationName: string): Location | null {
    try {
      const locConfig = this.locations.loadLocation(this.worldName, locationName)
      if (!locConfig) {
        logger.warning(`Location '${locationName}' not found in filesystem`)
        return null
      }

      const roomKey = RoomMappingService.locationToRoomKey(locationName)
      const existingAgents = this.rooms.getRoomMapping(this.worldName, roomKey)?.agents ?? []

      const dbLocation = createLocation(this.db, this.worldId, {
        name: locationName,
        displayName: locConfig.displayName,
        description: locConfig.description,
        positionX: locConfig.position[0],
        positionY: locConfig.position[1],
        adjacentTo: null,
        isDiscovered: locConfig.isDiscovered,
        // Keeps a not-yet-enriched location flagged as a draft on both sides.
        isDraft: locConfig.isDraft,
      })
      logger.info(`Created location '${locationName}' in database (id=${dbLocation.id})`)

      if (dbLocation.roomId) {
        this.rooms.setRoomMapping(this.worldName, roomKey, dbLocation.roomId, existingAgents)
        logger.info(`Stored room mapping: ${roomKey} -> room_id=${dbLocation.roomId}`)

        if (existingAgents.length > 0) {
          const added = this.addAgentsToRoom(dbLocation.roomId, existingAgents)
          logger.info(`Added ${added} character agents to room ${dbLocation.roomId}`)
        }
      }

      return dbLocation
    } catch (error) {
      logger.error(
        `Failed to create location '${locationName}' from filesystem: ${String(error)}`,
      )
      return null
    }
  }

  /**
   * Add agents to a room by name. The lookup is deliberately *not* scoped to
   * this world: narrowing it would change which agent ids land in `room_agents`
   * for existing databases where two worlds share a character name.
   */
  private addAgentsToRoom(roomId: number, agentNames: string[]): number {
    let added = 0
    for (const agentName of agentNames) {
      const agent = getAgentByName(this.db, agentName)
      if (!agent) {
        logger.warning(`Agent '${agentName}' not found in world '${this.worldName}'`)
        continue
      }
      addAgentToRoom(this.db, roomId, agent.id)
      added += 1
      logger.debug(`Added agent '${agentName}' to room ${roomId}`)
    }
    return added
  }

  // The same document goes to both sides, so the column holds `derived` too —
  // ignored by the clamping rules, returned by the API.
  saveStatDefinitions(statDefinitions: StatDefinitionsFile): void {
    this.players.saveStatDefinitions(this.worldName, statDefinitions)
    logger.info(`Saved stat definitions to filesystem for world '${this.worldName}'`)

    updateWorld(this.db, this.worldId, { statDefinitions: { ...statDefinitions } })
    logger.info(`Saved stat definitions to database for world_id=${this.worldId}`)
  }

  /**
   * Move the world to a new phase on both sides immediately; the deferred
   * variant is `WorldService.applyPendingPhase`. An unreadable `world.json`
   * still gets its row updated, or the player is stranded in onboarding.
   */
  updateWorldPhase(phase: WorldPhase): void {
    const config = this.worlds.loadWorldConfig(this.worldName)
    if (config) {
      config.phase = phase
      this.worlds.saveWorldConfig(this.worldName, config)
      logger.info(`Set phase='${phase}' in filesystem`)
    } else {
      logger.warning(
        `Could not load world config for '${this.worldName}' to update phase`,
      )
    }

    updateWorld(this.db, this.worldId, { phase })
    logger.info(`Set phase='${phase}' in database`)
  }

  /** Replace the database stat block from a `{name: value}` map. */
  syncStats(stats: Record<string, number>): void {
    initializePlayerStats(this.db, this.worldId, statDefinitionsFrom(stats), stats)
    logger.info(`Synced ${Object.keys(stats).length} stats to database`)
  }

  /**
   * Write the live database state back over the world's files, for backups and
   * portable exports. `effects`, `recent_actions`, `game_time`, `equipment` and
   * `flags` have no column and round-trip through the existing `player.json` —
   * exporting into a world missing that file silently resets the clock to day 1.
   * Locations export discovery status only.
   */
  exportStateToFilesystem(): void {
    const playerState = getPlayerState(this.db, this.worldId)
    const locations = getLocations(this.db, this.worldId)

    if (!playerState) {
      logger.warning(`No player state to export for world_id=${this.worldId}`)
      return
    }

    const fsState = this.players.loadPlayerState(this.worldName) ?? blankPlayerState()

    fsState.turnCount = playerState.turnCount ?? 0

    // Bare `JSON.parse` on purpose: a corrupt blob aborts the export rather
    // than overwriting a good `player.json` with nothing.
    if (playerState.stats) {
      fsState.stats = JSON.parse(playerState.stats) as Record<string, number>
    }
    if (playerState.inventory) {
      fsState.inventory = JSON.parse(playerState.inventory) as InventoryEntry[]
    }

    if (playerState.currentLocation) {
      fsState.currentLocation = playerState.currentLocation.name
    }

    this.players.savePlayerState(this.worldName, fsState)
    logger.info(`Exported player state to filesystem for world '${this.worldName}'`)

    for (const location of locations) {
      const locConfig = this.locations.loadLocation(this.worldName, location.name)
      if (!locConfig) continue
      // Only when discovery differs, so a backup does not rewrite every entry.
      if (locConfig.isDiscovered === Boolean(location.isDiscovered)) continue

      this.locations.updateLocation(this.worldName, location.name, {
        isDiscovered: location.isDiscovered,
        // `null` means "leave alone" to the writer.
        label: location.label,
      })
    }

    logger.info(`Exported ${locations.length} locations to filesystem`)
  }
}
